src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceService.cs

```diff
@@ -0,0 +1,294 @@
+
+            // If no namespace is specified, we just find and return the full NameSyntax.
+            if (old == nameRef || newNamespaceParts.IsDefaultOrEmpty)
+            {
+                @new = old;
+            }
+            else
+            {
+                if (newNamespaceParts.Length == 1 && newNamespaceParts[0].Length == 0)
+                {
+                    // If new namespace is "", then name will be declared in global namespace.
+                    // We will replace qualified reference with simple name + global alias.
+                    @new = SyntaxFactory.AliasQualifiedName(
+                        SyntaxFactory.IdentifierName(SyntaxFactory.Token(SyntaxKind.GlobalKeyword)), 
+                        nameRef.WithoutTrivia());
+                }
+                else
+                {
+                    var aliasQualifier = GetAliasQualifierOpt(old);
+                    var qualifiedNamespaceName = CreateNameSyntax(newNamespaceParts, aliasQualifier, newNamespaceParts.Length - 1);
+                    @new = SyntaxFactory.QualifiedName(qualifiedNamespaceName, nameRef.WithoutTrivia());
+                }
+
+                // We might lose some trivia associated with children of `outerMostNode`.  
+                @new = @new.WithTriviaFrom(old);
+            }
+            return true;
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
+
+        protected override async Task<(bool, NamespaceDeclarationSyntax)> ShouldPositionTriggerRefactoringAsync(
+            Document document, 
+            int position, 
+            CancellationToken cancellationToken)
+        {
+            var compilationUnit = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false) as CompilationUnitSyntax;            
+            var namespaceDecls = compilationUnit.DescendantNodes().OfType<NamespaceDeclarationSyntax>().ToImmutableArray();
+
+            // Here's conditions that trigger the refactoring (all have to be true in each scenario):
+            // 
+            // - There's only one namespace declaration in the document and all types are declared in it:
+            //    1. No nested namespace declaration (even it's empty).
+            //    2. The cursor is on the name of the namespace declaration.
+            //    3. The name of the namespace is valid (i.e. no errors).
+            //    4. No partial type declared in the namespace. Otherwise its multiple declaration will
+            //       end up in different namespace.
+            //
+            // - There's no namespace declaration and all types in the document are declared in global namespace:
+            //    1. The cursor is on the name of first declared type.
+            //    2. No partial type declared in the document. Otherwise its multiple declaration will
+            //       end up in different namespace.
```

Comment:
fwiw: and i think this can be done later, i think we shoudl also offer things at pos-0 in the file (like we offer "add file banner").

Reply:
Will bring this up during design meeting.